### DISCUSSION TITLE
Onboard References and Gamepad Signal

### DISCUSSION THREAD

**Discussion by FrancisGoncalves** [bitcraze/discussions#769] [Q&A]
Hello! I am working with crazyflies on my research, I have programmed my own controller on the firmware and until now I have been sending references from the computer with Python. My idea, now, is to program the references onboard in C and use a gamepad to make the crazyflie start following those references (basically launch it) by pressing one button. I have been digging and I found some of Arnaud's old code 'sequence.c' that I think will help me with the onboard references but after reading a lot of stuff I am still a bit lost on how to change the gamepad mapping in order to send this kind of 'activation' signal with one of the buttons.

Thank you!
Francisco

**Comment by ataffanel** [comment]
Hi,

I would need a little bit more precision in order to help.

Where is this code located, in what repos (maybe you can link to it)?

What would your setup look like? Do you have the game-pad connected to a PC or are you intending to do something else?

/Arnaud

**Reply by FrancisGoncalves** [reply]
The code is located [here](https://forum.bitcraze.io/viewtopic.php?f=6&t=2648&p=13352&hilit=demo#p13352). 
Yes, I have the gamepad connected to the PC, and I am using Vicon for position feedback. I want to be able to connect to the crazyflie through the client and just press a button to start following the onboard references. Let me know if you need anything else.

Thanks

**Reply by ataffanel** [reply]
Thanks for detailing your use-case, I do not think this example is a good starting point: it was done a long time ago before we had support for apps and out of tree builds.

I think a much better example is the swarm demo we have in the main firmware repos: https://github.com/bitcraze/crazyflie-firmware/tree/master/examples/demos/swarm_demo.

This is the firmware-part of a demo we use to run at conference's exhibition where a PC software, we call the control-tower, was instructing Crazyflies to take-off. The way this was done was by having trajectories and a state-machine in the Crazyflie and having the PC to set a parameter to start takeoff. All Crazyflie where standing and coming back to Qi charging landing pad, this made the demo autonomous.

In the code, the [takeoff](https://github.com/bitcraze/crazyflie-firmware/blob/2cf04d7099f96d3e37082a6e25df6055018a5e4a/examples/demos/swarm_demo/app.c#L459) parameter sets the `takeOffWhenReady` variable and then, in the state machine, we look for this variable to decide [if we can go further and take-off.](https://github.com/bitcraze/crazyflie-firmware/blob/2cf04d7099f96d3e37082a6e25df6055018a5e4a/examples/demos/swarm_demo/app.c#L261).

On the PC side, we can simply [set the takeoff parameter to 1](https://github.com/bitcraze/crazyflie-firmware/blob/2cf04d7099f96d3e37082a6e25df6055018a5e4a/examples/demos/swarm_demo/control_tower/control_tower.py#L127) when it is time to take-off. In your context this could be done when the gamepad button press is detected.

In your case though, I am assuming you are using ROS in order to fly with vicon. This mean that the Crazyflie-side of thing will be pretty similar the swarm demo (you might even be able to use it as-is, the only thing to change will be to make sure the charging handling in the state machine does not block you), you will have to adapt the PC side to read the gamepad and set parameters from ROS.
